src/match.c

```diff
@@ -1626,25 +1669,58 @@ r_obj* expand_compact_indices(const int* v_o_haystack,
     v_o_loc_needles = r_int_cbegin(o_loc_needles);
   }
 
-  const bool retain_remaining_haystack = (remaining->action != VCTRS_REMAINING_ACTION_drop);
-  int* v_detect_remaining_haystack = NULL;
-  if (retain_remaining_haystack) {
-    r_obj* detect_remaining_haystack = KEEP_N(r_alloc_integer(size_haystack), &n_prot);
-    v_detect_remaining_haystack = r_int_begin(detect_remaining_haystack);
+  bool any_multiple_needles = false;
+  bool any_multiple_haystack = false;
 
-    for (r_ssize i = 0; i < size_haystack; ++i) {
-      // Initialize to "remaining" (i.e. this haystack value wasn't matched)
-      v_detect_remaining_haystack[i] = 1;
-    }
-  }
+  r_ssize loc_first_multiple_needles = -1;
+  r_ssize loc_first_multiple_haystack = -1;
 
-  bool any_multiple = false;
-  bool check_for_multiple =
+  // Check is always needed for `multiple = "all"`.
+  // This also handles `relationship` options too, since if `multiple` is
+  // `"any"`, `"first"`, or `"last"`, we can't invalidate a `relationship`.
+  bool check_multiple_needles =
     multiple == VCTRS_MULTIPLE_all ||
+    // TODO: Remove deprecated support for `multiple = "error"/"warning"`
     multiple == VCTRS_MULTIPLE_error ||
     multiple == VCTRS_MULTIPLE_warning;
```

Comment:
Figuring out if we need to check for multiple matches in `needles` is much simpler than for `haystack`.

We basically just need to check if `multiple = "all"`, regardless of the value of `relationship`.

If `multiple = "all"`, we _have_ to check for multiple matches in `needles` no matter what because we do a sort at the end if we detect multiple matches.

If `multiple %in% c("any", "first", "last")` then we _don't_ need to check for multiple matches in `needles`, because by definition they will be reduced down to 1 match through `multiple`.